### PR TITLE
Improve/refactor completions

### PR DESCRIPTION
**Pull request by bioball**
* Add CompletionProvider interface
* Add providers for module uris, qualified access, and string literals
* Introduce ResolveVisitor for completion items
* Refactor qualified access comleter to use ResolveVisitors
* Add more trigger characters to trigger completion
* Add support for resolving more completion item details for module URIs
* Move documentation rendering to its own package

### CODE REVIEW

File: src/main/kotlin/org/pkl/lsp/packages/dto/PackageAssetUri.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+data class PackageAssetUri(val packageUri: PackageUri, val assetPath: String) {
+  companion object {
+    fun create(str: String): PackageAssetUri? {
+      val uri =
+        try {
+          URI(str)
+        } catch (e: URISyntaxException) {
+          return null
+        }
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/packages/dto/PackageAssetUri.kt:29]: Do you know about `runCatching`?
```suggestion
      val uri = runCatching { URI(str) }.getOrNull() ?: return null
```

File: src/main/kotlin/org/pkl/lsp/completion/StringLiteralTypeCompletionProvider.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  ) {
+    val stringLiteral = getStringLiteral(node) ?: return
+    val context = node.containingFile.pklProject
+    val resultType = stringLiteral.inferExprTypeFromContext(project.pklBaseModule, mapOf(), context)
+    addCompletionResults(resultType, stringLiteral, collector, null, project.pklBaseModule, context)
+  }
+
+  // only handle completions for single line string literals (multiline string literals are harder
+  // to insert, not
+  // likely to be used for enum members
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/completion/StringLiteralTypeCompletionProvider.kt:45]: nit
```suggestion
  // to insert, not likely to be used for enum members
```

File: src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+fun Type.toMarkdown(project: Project, context: PklProject?): String {
+  val markdown = render()
+  val ctx = getNode(project, context)
+  showDocCommentAndModule(ctx, markdown, context)
+  return when {
+    ctx is PklModule && ctx.header != null ->
+      showDocCommentAndModule(ctx.header!!, markdown, context)
+    else -> showDocCommentAndModule(ctx, markdown, context)
+  }
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt:44]: That `when` seems more complicated than it needs to be. More important, though, is the `showDocCommentAndModule(ctx` duplication (when `ctx` is _not_ `PklModule`) intentional?
```suggestion
    return showDocCommentAndModule((ctx as? PklModule)?.header ?: ctx, markdown, context)
```

File: src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+      }
+    is PklStringLiteralType -> "\"$text\""
+    is PklMethod -> {
+      val modifiers = modifiers.render()
+      modifiers + methodHeader.doRenderMarkdown(originalNode, context)
+    }
+    is PklMethodHeader ->
+      buildString {
+        append("function ")
+        append(identifier?.text ?: "<method>>")
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt:68]:
```suggestion
        append(identifier?.text ?: "<method>")
```

?

File: src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            returnType!!.doRenderMarkdown(originalNode, context)
+          } else {
+            val parent = this@doRenderMarkdown.parent
+            if (parent != null && parent is PklMethod) {
+              val type = parent.body.computeExprType(project.pklBaseModule, mapOf(), context)
+              type.render()
+            } else "unknown"
+          }
+        append(": ")
+        append(returnTypeStr)
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt:82]:
```suggestion
        append(": ")
        append(
          returnType?.doRenderMarkdown(originalNode, context) ?:
          (this@doRenderMarkdown.parent as? PklMethod)?.body?.computeExprType(project.pklBaseModule, mapOf(), context) ?:
          "unknown"
        )
```
?

File: src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        append(identifier?.text)
+        typeParameterList?.let { append(it.doRenderMarkdown(originalNode, context)) }
+      }
+    is PklType -> render()
+    else -> text
+  }
+
+// render modifiers
+private fun List<Terminal>?.render(): String {
+  return this?.let { if (isEmpty()) "" else joinToString(" ", postfix = " ") { it.text } } ?: ""
```
Comment (holzensp) [inline src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt:153]:
```suggestion
  return this?.let { joinToString(" ", postfix = " ") { it.text } } ?: ""
```

joining `listOf()` is `""`

File: src/main/kotlin/org/pkl/lsp/ast/Extensions.kt [hunk 7]
```diff
@@ -467,3 +440,47 @@ fun Node.utfAwareText(): String {
...
+private val quoteCharacters =
+  EnumSet.of(TokenType.SLQuote, TokenType.SLEndQuote, TokenType.MLQuote, TokenType.MLEndQuote)
+
+fun PklStringConstant.contentsSpan(): Span {
+  val characters =
+    terminals
+      .dropWhile { quoteCharacters.contains(it.type) }
+      .dropLastWhile { quoteCharacters.contains(it.type) }
+  val firstSpan = characters.first().span
+  val lastSpan = characters.last().span
```
[the remark this answers is not in the document]
Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/ast/Extensions.kt:453]: Good catch!

File: src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt (hunk 4 above)
Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt:68]: Good catch!

File: src/main/kotlin/org/pkl/lsp/type/Types.kt [hunk 8]
```diff
@@ -989,6 +991,11 @@ sealed class Type(val constraints: List<ConstraintExpr> = listOf()) {
       rightType.render(builder, nameRenderer)
     }
 
+    fun eachElementType(processor: (Type) -> Unit) {
```
[the remark this answers is not in the document]
Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/type/Types.kt:994]: Maybe `eachMemberType` is a better name. That's what we call them in `pkl.reflect#UnionType`.

**Pull request merged**